pset5/speller/dictionary.c

```diff
@@ -10,88 +10,114 @@
 
 // Create the node data type for the hash table
 // each element of the hash table array is a node pointer
 typedef struct node {
-    char word[LENGTH + 1];
-    struct node *next;
+    bool is_word;
+    struct node *children[CHAR_AMOUNT];
 } node;
 
-// Declare the hash table
-node *hashtable[HASH_SIZE] = {NULL};
+// Root of trie
+node *root;
 
-// TODO Make a hashtable datatype
 
 /**
  * Returns true if word is in dictionary else false.
  */
 bool check(const char *word)
 {
-    // Traverse the linked list in the hash table to find our word.
-    // Head of the linked list in the bucket we want to look through
-    node *head = hashtable[adler32(word, sizeof(word))];
+    // for each letter in input word
+      // go to corresponding element in children
+        // if NULL, word is mispelled
+        // if not NULL, move to next letter
+      // once at end of input word
+        // check if is_word true
     
-    node *cursor = head;
+    node *cursor = root;
+    int i;
     
-    while(cursor != NULL) {
-        
-        int comp = strcasecmp(cursor -> word, word);
-        
-        // if they match
-        if (comp == 0) {
-            // word is spelled correctly (in dictionary)
-            return true;
+    for (i = 0; i < strlen(word); i++) {
+        int index = letter_index(word[i]);
+        if (cursor -> children[index] == NULL) {
+            return false; 
         }
-        // if they don't match
         else {
-            // check next word
-            cursor = cursor -> next;
+            cursor = cursor -> children[index];
         }
     }
-    // If we are out here, that means we did not find the word or it
-    // was misspelled.
-    return false;
+    if (cursor -> is_word) {
+        return true;
+    }
+    else {
+        return false;
+    }
 }
 
+/* Finds a letter's place in the alphabet */
+int letter_index(char letter) {
+    int letter_index;
+    if (isupper(letter)) {
+        letter_index = letter % 65;
+    }
+    else if (islower(letter)) {
+        letter_index = letter % 97;
+    }
+    else {
+        // we have another char
+        letter_index = 27;
+    }
+    return letter_index;         
+}
 
-// Hash function for load
-int hash(const char *word) {
-    int index = 0;
+void insert_word(node *n, char *word) {
+    // Takes a pointer to a node and a string and it inserts
+    // the string into trie starting at that node
     int i;
     for (i = 0; i < strlen(word); i++) {
-        index += tolower(word[i]);
-    }
-    index = index % HASH_SIZE;
-    return index;  
-};
+        // get the letter's index in the alphabet
+        int index = letter_index(word[i]);
+
+        // each element in children corresponds to
+        // a different letter
+        // check the value at children[i]
+        if (n -> children[index] == NULL) {
+            // if NULL, malloc a new node, have 
+            // children[i] point to it
+            node *new_node = malloc(sizeof(node));
```

Comment:
it's fine to have an intermediate variable here, but I think it's actually harder to follow than not.

`n->children[index] = malloc(sizeof(node));`